### ISSUE TITLE
Parallelize tests cause multiple deploys

### ISSUE THREAD

**Issue by davidrv** [Shippable/support#4298]
Hi, we have parallelized tests for our integration in Shippable. We also have a deploy to Heroku configured in the `PostCi` section of the `shippable.yml` file.

The problem is that, every time a single individual test suite succeeds, the deploy is performed. Even if the other test suits of the matrix are still `processing` or `pending`. 

How can we perform only one single deploy right after (and only after) all the parallelized tests suits are succeeded? 

- https://app.shippable.com/github/localistico/api/runs/64/summary/console

Attached a reduced size version of our `shippable.yml` file...

[shippable.yml.zip](https://github.com/Shippable/support/files/1954830/shippable.yml.zip)

**Comment by immerrr** [comment]
FTR, I've succeeded in adding a separate deploy step for the repo in question, so that it runs after all matrix builds succeed, but this is way too complex. It took me good few hours to understand the intricacies of workflow jobs and syncrepos, things that we could probably live without just fine IF there was a "on_success" callback that runs after all matrix builds. I mean, a downstream job is triggered from somewhere, and it's only triggered once, so there's clearly a place where a certain build step could be executed.

**Comment by trriplejay** [comment]
Looks like you found the right way to implement it. Unfortunately we do not currently have any kind of final callback section for matrix builds.

We've had this request come up a few times and i don't think this support is on our roadmap.  Ideally we'd instead like to improve the assembly line experience to make this scenario more straightforward to implement via jobs and resources, so that it comes easily and you wouldnt have to spend hours figuring it out.

https://github.com/Shippable/support/issues/1493
https://github.com/Shippable/support/issues/1592

I think the main issue is that the matrix jobs all can run on different physical machines, so any final callback would have to lose the context of the individual jobs.  The runSh lets you set up the context again as you see fit, though as you've seen it comes with its own challenges and solutions that are not obvious.

**Comment by immerrr** [comment]
> I think the main issue is that the matrix jobs all can run on different physical machines, so any final callback would have to lose the context of the individual jobs. 

FWIW, in this case I would be just fine to lose whatever context except for two pieces of information: a branch and a commit for which all the tests succeeded. For example, it would be just fine if there was a way via shipctl or anything to check if all other sibling matrix builds completed successfully, then one could do something like this
```
on_success:
  - if shippable_ctl sibling_matrix_builds_succeeded; then do_deploy.sh; done
```
and avoid treading the "workflow" path entirely.

I'm not saying workflows are necessarily bad, they are very flexible and we would probably get there naturally, since with docker the workflow quite naturally arranges itself into "build image -> test image (N matrix builds) -> deploy image" steps. I'm saying it's probably not that necessary if all you need to do is to collect statuses of all matrix builds.

Also, it would be nice to have a complete step-by-step guide on how to do that, because I had to thread together information from quite a few sources, including responses in this exact repository. There are quite a few peculiar moving parts, like pointing to a specific branch to set up the syncrepo, because if you're doing it the github way via a PR, the master won't receive the pipeline definition until the PR is merged.

**Comment by immerrr** [comment]
There's one more question that seems important and I still don't know the answer: is there a possibility of a race condition when two concurrent workflow step invocations with matrix builds of their own create a IN-OUT resource for the next workflow step they trigger?

Suppose we have two steps in the workflow, `A` and `B`, suppose there's resource `R` produced as OUT in `A` and consumed as IN in `B`. Suppose then that `A` has 3 matrix builds, `A-1`, `A-2` and `A-3`. As far as I can see in the UI, each of those builds upon completion will create its own copy of `R` (so one has to use `switch: off` in `B`'s resource definition).

Is it possible for the following sequence of events to happen?
```
- A1 is triggered by a commit
- A1-1 runs, creates R1-1
- A2 is triggered by a commit
- A1-2 runs, creates R1-2
- A2-1 runs, creates R2-1
- A1-3 runs, creates R1-3, triggers B1
- A2-2 finishes, creates R2-2
- B1 runs, consumes A1 and version R2-2 of resource R
```

Or is there a guarantee that if `A1-3` is the last matrix build of a step that triggers a subsequent workflow step, that workflow step will only consume the `R1-3` version of the resource?

**Comment by immerrr** [comment]
One more question: I ended up with two "assembly line" configurations, one for master and one for the feature branch in which i enabled the pipeline. How do I remove the second one? I can only see buttons to add or edit existing assembly lines.

**Comment by trriplejay** [comment]
**deleting your old branch**

In the SPOG, (or in the grid-view, make sure the resources table is visible) you can find the syncRepo that is attached to the rSync job.  Right click on it, then select "remove".  Make sure you choose the syncRepo for the correct branch. The branch is typically part of the resource name.  Removing a syncRepo will remove all of the resources and jobs that it created.  This performs a kind of "soft" delete of everything.  Now you can find those resources in the "deleted objects" table.  If you find your syncRepo there again, you can delete it again.  This time, the delete will be permanent and will entirely remove all associated resources from your subscription.

**race condition**

in short, based on your scenario, yes it is possible that an A1 job completing could trigger a deployment of something that A2 is in the middle of validating. Even without the intermediate resource, I think the main issue is having multiple jobs running in parallel. We can't predict which ones will finish first, especially since they might be running on dynamic nodes in various states.

I would say that you can reduce this possibility by removing the intermediate gitRepo OUT step you have.  Instead of doing a gitRepo OUT with replicate, you can write the commit sha directly to the job state.  Then, in your subsequent job (where you still have the gitRepo IN), you can use that commit sha to do a `git checkout` in your repo's state directory.  Your gitRepo would already hold the newer commits due to webhook processing, so you should just be able to check out the commit you're about to deploy.

This solves the case where A1 completing deploys A2, since B is triggered directly by A, and reads A's latest state in the moment that it is triggered.  However, it is still possible for the overall A2 run to finish before the overall A1 run, thus deploying A2's result before A1's result.  This can only be fixed by placing a limit on parallel jobs. 

**dependencyModes**
One other thing you could try that would be a bit simpler...  keep the gitRepo OUT, but don't switch it off. use `buildOnCommit: false` to prevent it from updating on webhooks.  Remove the direct job to job connection, and instead add a `dependencyMode: strict` to your deploy job (documented [here](http://docs.shippable.com/platform/workflow/overview/#trigger-modes)).  The result would be that deploy is only triggered when *all* of the incoming jobs are in a final state.  So whichever job finishes processing last gets deployed.

so, in your same flow listed above, job B would only run once, and it would only run after both A1 and A2 have completed, and it will deploy whatever the latest/most-recent version of the gitRepo is.

The resulting yml would look something like this:
```yml
resources:
  - name: myRepo
    type: gitRepo
    integration: myGithub
    versionTemplate:
      sourceName: trriplejay/simpleserver
      buildOnCommit: false

jobs:
  - name: deployChanges
    type: runSh
    dependencyMode: strict
    steps:
      - IN: myRepo
      - TASK:
        - script: echo "deploying now"

  - name: simpleserver_runCI
    type: runCI
    steps:
      - OUT: myRepo
        replicate: simpleserver_ciRepo

```

I hope that helps. happy to answer any additional questions you have.  I'll try to turn this into a sample for our docs.
